databuilder/extractor/delta_lake_metadata_extractor.py

```diff
@@ -0,0 +1,316 @@
+        self.spark = spark
+
+    def extract(self) -> Union[TableMetadata, TableLastUpdated, None]:
+        if not self._extract_iter:
+            self._extract_iter = self._get_extract_iter()
+        try:
+            return next(self._extract_iter)
+        except StopIteration:
+            return None
+
+    def get_scope(self) -> str:
+        return 'extractor.delta_lake_table_metadata'
+
+    def _get_extract_iter(self) -> Iterator[Union[TableMetadata, TableLastUpdated, None]]:
+        """
+        Given either a list of schemas, or a list of exclude schemas,
+        it will query hive metastore and then access delta log
+        to get all of the metadata for your delta tables. It will produce:
+         - table and column metadata
+         - last updated information
+        """
+        if self.schema_list:
+            LOGGER.info("working on {}".format(self.schema_list))
+            tables = self.get_all_tables(self.schema_list)
+        else:
+            LOGGER.info("fetching all schemas")
+            LOGGER.info("Excluding: {}".format(self.exclude_list))
+            schemas = self.get_schemas(self.exclude_list)
+            LOGGER.info("working on {}".format(schemas))
+            tables = self.get_all_tables(schemas)
+        # TODO add the programmatic information as well?
+        # TODO add watermarks
+        for table in tables:
+            LOGGER.info("scraping {}".format(table.name))
+            scraped_table = self.scrape_table(table)
+            if self.delta_tables_only and not scraped_table.is_delta_table():
+                LOGGER.info("Skipping none delta table {}".format(table.name))
+                continue
+            else:
+                yield self.create_table_metadata(scraped_table)
+                last_updated = self.create_table_last_updated(scraped_table)
+                if last_updated:
+                    yield last_updated
+
+    def get_schemas(self, exclude_list: List[str]) -> List[str]:
+        '''Returns all schemas.'''
+        schemas = self.spark.catalog.listDatabases()
+        ret = []
+        for schema in schemas:
+            if schema.name not in exclude_list:
+                ret.append(schema.name)
+        return ret
+
+    def get_all_tables(self, schemas: List[str]) -> List[Table]:
+        '''Returns all tables.'''
+        ret = []
+        for schema in schemas:
+            ret.extend(self.get_tables_for_schema(schema))
+        return ret
+
+    def get_tables_for_schema(self, schema: str) -> List[Table]:
+        '''Returns all tables for a specific schema.'''
+        return self.spark.catalog.listTables(schema)
+
+    def scrape_all_tables(self, tables: List[Table]) -> List[ScrapedTableMetadata]:
+        return [self.scrape_table(table) for table in tables]
+
+    def scrape_table(self, table: Table) -> ScrapedTableMetadata:
+        '''Takes a table object and creates a scraped table metadata object.'''
```

Comment:
could you share more context on the scrape table metadata object? is it because it doesn't include all metadata for the full table metadata?

Reply:
similar problem as with ColumnMetadata. I am trying to batch getting all of the possible metadata that I can from delta-lake tables at the same time and not all parts of the metadata have an amundsen equivalent yet. It seemed like at this point in amundsen development at least, it would be easier to just scrape it and store it in a custom python class and then map to the amundsen specific classes later. For example, a ScrapedTableMetadata can actually produce a TableMetadata, LastUpdatedMetadata etc. 

I can add todos saying that it would be good to revisit this though.